### ISSUE TITLE
integrate.solve_bvp does not check for convergence of boundary conditions

### ISSUE THREAD

**Issue by scbarton** [scipy/scipy#9832]
I’d like to report an issue with `scipy.integrate.solve_bvp`. This is an excellent piece of code but it unfortunately does not check for convergence of boundary conditions. I can demonstrate with the following example code with nonlinear boundary conditions. The code solves Laplace's equation, which is linear and converges in only one iteration.

### Reproducing code example:
```
import numpy as np
from scipy.integrate import solve_bvp

# Parameters
kappa= 1.64
ioA, ioC= 0.01, 1.0e-4
V, UA, UC= 0.5, 0.0, 1.0
f= 38.9
L=0.1

## Governing Equation- Laplace Eq. for potential
def deq(x,y):
    return np.stack([y[1], np.zeros_like(x)])

## Boundary Conditions
def bc(uA,uC):

    phiA, phipA= uA
    phiC, phipC= uC

    # Butler-Volmer Kinetics at Anode
    hA=0-phiA-0
    iA= ioA * ( np.exp(f*hA) - np.exp(-f*hA))
    res0= iA + kappa*phipA

    # Butler-Volmer Kinetics at Cathode
    hC= V - phiC - UC
    iC= ioC * (np.exp(f*hC) - np.exp(-f*hC))
    res1= iC - kappa*phipC

    return np.array([  res0, res1 ] )
        
# Initial Guess
xinit=np.linspace(0,L)
uinit=np.array([0,0]) [:,None]   # a column vector
yinit=uinit * np.ones([len(uinit),len(xinit)])

# Find Solution
sol = solve_bvp(deq, bc, xinit, yinit, verbose=2)

# Calculate Boundary residuals
bc_res = bc(sol.y[:, 0], sol.y[:, -1])
print('Boundary Condition Residuals: ',bc_res)```
```

Using scipy v1.2.0, the above code generates the following output:

```
   Iteration    Max residual    Total nodes    Nodes added  
       1          3.39e-14          50              0       
Solved in 1 iterations, number of nodes 50, maximum relative residual 3.39e-14.
Boundary Condition Residuals:  [ 5.25230866e-02 -2.30274472e+02]
```

After only one iteration, the BC residuals are much greater than the tolerance (the default tolerance, 1e-3).

### Proposed solution
I have added code to check for convergence, modifying lines 1081-1111 of `_bvp.py` as follows:

```
        bc_res = bc_wrapped(y[:, 0], y[:, -1], p)  
        
        # This relation is not trivial, but can be verified.
        r_middle = 1.5 * col_res / h
        sol = create_spline(y, f, x, h)
        rms_res = estimate_rms_residuals(fun_wrapped, sol, x, h, p,
                                         r_middle, f_middle)
        max_rms_res = np.max(rms_res)
        max_bc_res = np.max(np.abs(bc_res))
        max_res = np.max([max_rms_res, max_bc_res])
		
        if singular:
            status = 2
            break

        insert_1, = np.nonzero((rms_res > tol) & (rms_res < 100 * tol))
        insert_2, = np.nonzero(rms_res >= 100 * tol)
        nodes_added = insert_1.shape[0] + 2 * insert_2.shape[0]

        if m + nodes_added > max_nodes:
            status = 1
            if verbose == 2:
                nodes_added = "({})".format(nodes_added)
                print_iteration_progress(iteration, max_res, m,
                                         nodes_added)
            break

        if verbose == 2:
            print_iteration_progress(iteration, max_res, m, nodes_added)

        if nodes_added > 0:
            x = modify_mesh(x, insert_1, insert_2)
            h = np.diff(x)
            y = sol(x)
        elif max_res <= tol:
            status = 0
            break
```

This code calculates `max_res` for both the domain and the boundary, and only breaks the parent `while` loop when that `max_res` falls below `tol`.

Using this version of `solve_bvp`, the same Laplace Equation code results in multiple iterations to reduce the boundary residuals to less than the tolerance.

```
   Iteration    Max residual    Total nodes    Nodes added  
       1          2.30e+02          50              0       
       2          2.86e+00          50              0       
       3          6.18e-03          50              0       
       4          7.25e-06          50              0       
Solved in 4 iterations, number of nodes 50, maximum relative residual 7.25e-06.
Boundary Condition Residuals:  [ 2.04326239e-07 -7.25180010e-06]

```

I've attached the modified verion of `_bvp.py`, and I'll be glad to make a pull request if needed.

### Scipy/Numpy/Python version information:

```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)

1.2.0 1.16.1 sys.version_info(major=3, minor=7, micro=2, releaselevel='final', serial=0)
```
[_bvp-scb.py.txt](https://github.com/scipy/scipy/files/2873413/_bvp-scb.py.txt)

**Comment by nmayorov** [comment]
Thanks for the meaningful report!

I think your idea is correct, especially because, as I figured, you work with bvp's often in your professional field. 

I would suggest the following modification --- introduce another parameter `bc_tol=None`. I think this is more correct as we use `tol` for relative residuals tolerance and it doesn't look appropriate to compare absolute boundary condition residuals with the same number. Additionally we will be able to choose appropriate different tolerance or disable the check by using `bc_tol=None`.

Could you please consider this and make a pull request?

**Comment by scbarton** [comment]
Thanks for your response.  I suggest that, by default, the boundary residuals should fall below *some* tolerance before the solution is returned. If not, the result returned is not necessarily a solution to the bvp. Novice users may not notice this error.

It's easy to adjust the values of the boundary residuals by multiplication within the `bc` function, so I consider it reasonable to compare these residuals to `tol`. I don't consider the additional parameter 'bc_tol` to be necessary, but if it is to be added, I suggest that the value default to `tol`.

**Comment by nmayorov** [comment]
I see what you mean.

> It's easy to adjust the values of the boundary residuals by multiplication within the `bc` function

I think it is generally not the best approach to interface design of a solver, it makes it difficult to understand how to tune the solver properly (by adjusting the problem definition).

How do you like the following. We can measure the scale of the boundary condition residuals as their abs values at the initial guess (to handle situations near zero and consider `atol`\`rtol` simultaneously). The we check `abs(bc) < tol * bc_scale`, which looks more correct to me. In this case we are probably OK without an additional `bc_tol` parameter.

**Comment by scbarton** [comment]
Something like this should work, and handles the parameters, `p`. It also passes all the tests.

```
        bc_res = bc_wrapped(y[:, 0], y[:, -1], p)

        # calculate boundary jacobian
        if bc_jac_wrapped is None:
            dbc=estimate_bc_jac(bc_wrapped,y[:, 0], y[:, -1], p)
        else:
            dbc=bc_jac_wrapped(y[:, 0], y[:, -1], p)
            
        #breakpoint()
        
        # stack the boundary jacobian with n+k columns, handling optional parameters
        if len(p)>0:
            ystack=np.vstack(dbc[0:2])
            pstack=np.vstack([dbc[2],np.zeros_like(dbc[2])])
            bcstack=np.hstack([ystack,pstack])
        else:
            bcstack=np.vstack(dbc[0:2])

        # calculate boundary residuals for y and p, handling zeros in boundary jacobian
        bc_yp_res=np.divide(bc_res, bcstack, out=np.zeros_like(bcstack), where=bcstack!=0 )
        max_bc_res=np.max(abs(bc_yp_res))
```
I feel like I'm reinventing the wheel though, isn't this done somewhere else in the code?

**Comment by nmayorov** [comment]
So here you are trying to divide residuals by each row of Jacobian matrices? Could you explain the mathematical logic behind this. 

Dividing something by its derivative doesn't create dimensionless quantity as far as I understand, so it doesn't look as a clean choice as a relative tolerance criteria.

**Comment by scbarton** [comment]
The idea is to convert the boundary residuals to residuals on the scale of the `y` matrix. In other words, `dy= df / (df/dy)`, where `df` is the boundary residual and `df/dy` is the boundary jacobian.  

No, these are not dimensionless. However, they could be further nondimensionalized in the same way as `y`.

**Comment by nmayorov** [comment]
> The idea is to convert the boundary residuals to residuals on the scale of the `y` matrix. In other words, `dy= df / (df/dy)`, where `df` is the boundary residual and `df/dy` is the boundary jacobian.

So it is something close to computing "how much do we need to change `y` to change boundary residuals by `df`". Am I correct?

I'm not sure about the divisions though. If we write it like this:
```
dbc = (dbc / dya) dya + (dbc / dyb) dyb + (dbc / dp) dp
```
Could we perhaps work from this formula? For `dya` we could substitute `dya = tol * (1 + abs(ya))` and then bound L1 norms `|(dbc / dya) dya|` by `|dbc / dya| |dya|` and do the same with other summands. Then finally check if `|bc_res| < |dbc / dya| |dya| + |dbc / dyb| |dyb| + |dbc / dp| |dp|`

What do you think?

**Comment by scbarton** [comment]
Yes, that is very straigthforward and avoids division by zero.  I will work on that.

On a side note, there does exist a `tol_bc` in the function `solve_newton` which is set kind of arbitrarily to `5e-2 * tol`.  I am not using that value, however, aside from adding `abs()` to line 493 where it is compared to `bc_res`.

**Comment by scbarton** [comment]
I've updated the pull request based on the above discussion.  HOWEVER I did not remove the division. Here is my logic:

We would like to calculate set of normalized boundary condition residuals `bc_res_norm`, to compare to `tol`, just like we compare `rms_res` to `tol`.  To do that, I calculate a denominator equal to 

```denom = (dbc / dya) * (1 + abs(ya)) + (dbc / dyb) * (1 + abs(yb)) + (dbc / dp) (1 + abs(p))  ```

This *should* always result in a full array (no zeros) with the same dimension as `bc_res`.  Then 
```bc_res_norm = bc_res / denom```
and we can compare the maximum value of this array to `tol`.  

What do you think? The main reason to do that is to produce a number comparable to `tol` that we can use to track convergence.  The approach that we previously discussed doesn't allow us to do that.

**Comment by nmayorov** [comment]
I'm getting a bit confused to be honest. 

Let's go to the basics. If we estimate some parameter `x` by a numerical method and we have an error estimate (or residual) `dx` we typically want to ensure that `abs(dx) < rtol * abs(x) + atol`. In our case we want to drive boundary conditions residuals to 0, so comparing `bc_res < tol` makes a good sense. 

What if I again suggest to introduce another tolerance parameter `bc_tol` (possibly equal to `1e-3`) and do the very basic check `bc_res < bc_tol`. I think it might be actually better and much easier to understand.

**Comment by scbarton** [comment]
Sure `bc_res < bc_tol` would be great, especially if `bc_tol` is an optional parameter that defaults to `tol`. 

This was the approach I used in my original revision. Everything since has been focused on normalizing `bc_res`, but I don't feel strongly about that.

Just for the record, `bc_res_norm ` represents `dy/(abs(y) + 1)`, so if we consider 

```rtol * abs(y) + atol = rtol * (abs(y) + 1)```

Then `abs(bc_res_norm) < rtol` is the same as `abs(dx) < rtol * abs(x) + atol`.

**Comment by nmayorov** [comment]
I think an ideal approach is to compare both strategies i.e.

1. `bc_res < bc_tol`
2. `bc_res_norm < tol * bc_scale(=denom)` 

And if we are to choose the 2nd (more sophisticated) strategy, we need to test that it doesn't prevent convergence and stuff like that. 

Could you test how reasonable the strategy 2 behaves at least on 3-4 problems?

**Comment by scbarton** [comment]
The second strategy passes all of the tests, which as you know includes several problems.  I did need to lower the tolerance on my new test, `test_nonlin_bc`, to be equal to the assertion tolerance of 1e-5 on `sol(x)`.

Do you think we need to test against more problems? I can think of a few.

**Comment by nmayorov** [comment]
If you think the first approach is better, I suggest to go with it. 

The second approach --- it something we invented and it might work well, but still it is something unconventional. And as it doesn't solve any fundamental problem, perhaps we should prefer a simpler approach.

So make the changes, I will review your pull request. I have a small qualitative benchmark of BVPs, I'm going to rerun it with your changes.

**Comment by scbarton** [comment]
The latest commit removes all normalization and implements the optional parameter `tol_bc`. In order to monitor convergence, I simply multiply the largest value of `bc_res` by `tol / tol_bc`.

**Comment by nmayorov** [comment]
Resolved by #9856
